### ISSUE TITLE
Cache warming?

### ISSUE THREAD

**Issue by dlai286** [spatie/laravel-responsecache#24]
Here is my scenario:

I run a product catalogue site that has around 20K individual pages cached on redis. The site gets around 500 to 1000 page requests per minute at any given time. Whenever I push an update, I have to flush the entire cache (around 1GB) which results in my server getting hit with a ton of uncached requests and ultimately kills the server. Updates have been a nightmare due to this.

Is there a way to build a "shadow" cache when I update then swap it with the real cache once done? My site does not have any user generated content, its basically a database driven catalogue for thousands of products (static pages).

I've tried to find the right cache lifetime but problems arise with Laravel's elixir asset bundle versioning (as the old bundle will no longer exist), I need to do entire cache swaps within 1 minute window to not aggravate users who hit stale pages that request old non-existing versioned bundles.

Basically something like:

0. Push update to server

1. `php artisan responsecache:warm`
2. Gets all `laravel-responsecache` keys and reverses hash for URL.
3. Generates a fresh cached copy and stores it to `'shadowCacheStore' => env('SHADOW_RESPONSE_CACHE_DRIVER', 'file'),`
4. Once all are done, swap real cache with shadow (such that site can be updated as instantly as possible).
5. Delete shadow cache.

Hope I'm making sense, I originally thought this was a common use case but I cannot find any related cache invalidation strategies (I just came up with the shadow cache method just as I was typing this).

Your package is a life saver so far, would be awesome if it could make site updates less painful!

**Comment by freekmurze** [comment]
Hi,

I'm glad you find the package useful. It's awesome that it is being used on such a high traffic site. Could you share which site that is?

Unfortunately there currently is no way to do cache warming. I believe the package alone is and should not be responsible for cache warming. What you could do is set up the site a second time on the same or differente server, crawl all pages with a crawler (so the cache is warmed up) and then switch from the old site to the new site. When you're doing updates again crawl the old site to warm it's cache and switch back to there.

I'm also making this up as I type. I think it would be best to ask some people that are more familiar with solving problems like this on high traffic sites. They will probably know a more elegant solution.

**Comment by viezel** [comment]
I know this thread is quite old, but why not use Varnish or any other reverse proxy for this. 
It has BAN tagging purge support, so we only have to clear then data that have been changed. 

Secondly, Varnish will be much faster than this package, since this still evaluates through php and your 1000 req/min can be done with one web server and one Varnish.

**Comment by freekmurze** [comment]
If you want to use Varnish, do so. We even have a package to make integration of Varnish easier in a Laravel app.

The benefit of this responsecache package is that you can easily install it in any app without having to install extra software on the server. You can configure this package using a simple config file and do not have to learn Varnish’ configuration syntax.
